crates/ra_mbe/src/lib.rs

```diff
@@ -50,6 +52,42 @@ pub(crate) struct Rule {
     pub(crate) rhs: tt::Subtree,
 }
 
+/// Find the "shift" (the highest id of the TokenId) inside a subtree
+fn find_subtree_shift(tt: &tt::Subtree, mut cur: Option<u32>) -> Option<u32> {
+    use std::cmp::max;
+
+    for t in &tt.token_trees {
+        cur = match t {
+            tt::TokenTree::Leaf(leaf) => match leaf {
+                tt::Leaf::Ident(ident) if ident.id != tt::TokenId::unspecified() => {
+                    Some(max(cur.unwrap_or(0), ident.id.0))
+                }
+                _ => cur,
+            },
+            tt::TokenTree::Subtree(tt) => find_subtree_shift(tt, cur),
+        }
+    }
+
+    cur
+}
```

Comment:
I think this will be slightly more clear if expresse in terms of stateless iterator:

```rust
fn max_id(tt: &tt::TokenTree) -> Option<u32> {
  match tt {
    tt::Subtree { token_trees, .. } => token_trees.iter().filter_map(max_id).max(),
    tt::Leaf => compute directly
  }
}
```

Reply:
Also, for the naming, I feel it's sligthly more precise to say that first, we are finding the highest `id`, and then turn it into a `shift`, rather than using the same word for both cases. The reason is that "find highest id" is an operation that makes sense even without refering to future shifring of trees.